File: android/app/src/main/java/com/gproust/sprout/ui/home/MedicineWatchCard.kt
package com.gproust.sprout.ui.home

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowRight
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedCard
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.pluralStringResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.gproust.sprout.R
import com.gproust.sprout.data.MedicineWatch
import com.gproust.sprout.data.local.MedicineEntity
import com.gproust.sprout.ui.medicines.levelColor
import com.gproust.sprout.ui.medicines.levelIcon
import com.gproust.sprout.ui.medicines.stateSentence

/**
 * How many medicines the dashboard will show before it stops being a glance.
 *
 * A household with more waits running than this has a screen for them, and the
 * card says how many it is not showing rather than growing to fit.
 */
private const val WATCH_LIMIT = 3

/**
 * The as-needed medicines with a wait running, or one that has just finished
 * (BDR-16).
 *
 * It is on the dashboard because that is where a parent already is at 3 a.m.,
 * and "has the paracetamol had six hours yet" is a question asked far more
 * often than it is answered by opening a second screen. It says nothing the
 * as-needed screen doesn't — same sentence, same icon, same colour, from the
 * same helpers — and it is absent entirely when no wait is running, so the
 * dashboard of a household that is not in the middle of anything is unchanged.
 *
 * *Give a dose* is offered here for the same reason it is never disabled
 * there: the dose that goes unlogged is the one the parent had to leave the
 * screen to record.
 */
@Composable
fun MedicineWatchCard(
    watches: List<MedicineWatch>,
    now: Long,
    onGive: (MedicineEntity) -> Unit,
    onOpen: () -> Unit,
    modifier: Modifier = Modifier,
) {
    if (watches.isEmpty()) return
    val context = LocalContext.current
    // Outlined rather than filled: on a household dashboard this sits *inside*
    // a baby's card, and two filled surfaces at the same tone would read as one.
    OutlinedCard(modifier.fillMaxWidth()) {
        Column(Modifier.padding(start = 16.dp, top = 12.dp, bottom = 12.dp, end = 4.dp)) {
            Row(
                Modifier.fillMaxWidth().clickable(onClick = onOpen).padding(end = 12.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(
                    stringResource(R.string.home_medicine_title),
                    style = MaterialTheme.typography.titleSmall,
                    fontWeight = FontWeight.SemiBold,
                    modifier = Modifier.weight(1f),
                )
                Icon(
                    Icons.AutoMirrored.Filled.KeyboardArrowRight,
                    contentDescription = stringResource(R.string.screen_medicines),
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }

            watches.take(WATCH_LIMIT).forEach { watch ->
                Spacer(Modifier.height(8.dp))
                WatchRow(
                    watch = watch,
                    sentence = stateSentence(context, watch.readiness, now),
                    onGive = { onGive(watch.medicine) },
                )
            }

            val hidden = watches.size - WATCH_LIMIT
            if (hidden > 0) {
                Text(
                    pluralStringResource(R.plurals.home_medicine_more, hidden, hidden),
                    style = MaterialTheme.typography.labelMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    modifier = Modifier.padding(top = 8.dp),
                )
            }
        }
    }
}

/**
 * One medicine's line: the state three ways over — a dot, its own icon and a
 * sentence — and the dose beside it.
 *
 * The three channels are not decoration. Red/amber/green is the palette a
 * deuteranope reads worst, and this card exists to be read at a glance by
 * someone frightened and half awake (BDR-15).
 */
@Composable
private fun WatchRow(watch: MedicineWatch, sentence: String, onGive: () -> Unit) {
    val colour = levelColor(watch.readiness.level)
    Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
        Spacer(
            Modifier
                .size(10.dp)
                .background(colour, CircleShape),
        )
        Spacer(Modifier.width(8.dp))
        Icon(
            levelIcon(watch.readiness.level),
            // Decorative: the sentence below says the same thing, and a screen
            // reader announcing both would say it twice.
            contentDescription = null,
            tint = colour,
            modifier = Modifier.size(18.dp),
        )
        Column(Modifier.weight(1f).padding(start = 8.dp)) {
            Text(watch.medicine.name, style = MaterialTheme.typography.bodyMedium)
            Text(sentence, style = MaterialTheme.typography.labelMedium, color = colour)
        }
        // Never disabled, whatever the light says — the same rule the as-needed
        // screen keeps. Sprout records what happened; it does not decide it.
        TextButton(onClick = onGive) {
            Text(stringResource(R.string.medicine_give))
        }
    }
}

File: ios/Sprout/Sources/Screens/Home/MedicineWatchCard.swift
import SproutData
import SwiftUI

/// How many medicines the dashboard will show before it stops being a glance.
///
/// A household with more waits running than this has a screen for them, and the
/// card says how many it is not showing rather than growing to fit.
private let watchLimit = 3

/// The as-needed medicines with a wait running, or one that has just finished
/// (BDR-16) — the mirror of `ui/home/MedicineWatchCard.kt`.
///
/// **One line each, and no more.** The first version of this said everything the
/// As needed screen says — name, dose, interval, the full state sentence, the
/// last dose and its count — and the first thing users said back was that the
/// dashboard had got heavy. They were right: this is a glance at something that
/// is usually not happening, sitting on the screen that is opened every hour. So
/// a line is the name in its state's colour, the number still to wait if there
/// is one, and the two things there are to do about it.
///
/// It keeps its place under the feed buttons rather than above them, for the
/// same reason: feeding is what the dashboard is opened for, and a medicine is
/// what it is opened for a few days a year.
///
/// Everything the line drops is a tap away on the As needed screen — and none of
/// it is dropped for VoiceOver, which is given the full sentence.
struct MedicineWatchCard: View {
    let watches: [MedicineWatch]
    let now: Int64
    let onGive: (MedicineWatch) -> Void
    let onDismiss: (MedicineWatch) -> Void
    let onOpen: () -> Void

    var body: some View {
        if watches.isEmpty {
            EmptyView()
        } else {
            VStack(spacing: 0) {
                ForEach(watches.prefix(watchLimit)) { watch in
                    WatchRow(
                        watch: watch,
                        now: now,
                        onOpen: onOpen,
                        onGive: { onGive(watch) },
                        onDismiss: { onDismiss(watch) }
                    )
                }

                if watches.count > watchLimit {
                    Button(action: onOpen) {
                        Text(Str.t("home_medicine_more", watches.count - watchLimit))
                            .font(.caption)
                            .foregroundStyle(SproutColor.onSurfaceVariant)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .padding(.horizontal, Spacing.snug)
                            .padding(.bottom, Spacing.tight)
                    }
                    .buttonStyle(.plain)
                }
            }
            .padding(.vertical, Spacing.hairline)
            .background(
                RoundedRectangle(cornerRadius: Radius.card)
                    .stroke(SproutColor.outline.opacity(0.4), lineWidth: 1)
            )
        }
    }
}

/// One medicine, one line: `⏳ Paracetamol  4 h 12 m to wait   Give  ✕`.
///
/// The state is carried three ways over — the symbol's shape, the colour, and
/// the words of the countdown — because red/amber/green is the palette a
/// deuteranope reads worst and this is read at 3 a.m. by someone frightened
/// (BDR-15). The symbol is what distinguishes *can be given* from *can be given,
/// sooner than ideal* once there is no number left to print.
private struct WatchRow: View {
    let watch: MedicineWatch
    let now: Int64
    let onOpen: () -> Void
    let onGive: () -> Void
    let onDismiss: () -> Void

    private var colour: Color { watch.readiness.level.color }

    var body: some View {
        HStack(spacing: Spacing.tight) {
            Button(action: onOpen) {
                HStack(spacing: Spacing.tight) {
                    Image(systemName: watch.readiness.level.symbol)
                        .font(.footnote)
                        .foregroundStyle(colour)
                    Text(watch.medicine.name)
                        .font(.callout.weight(.medium))
                        .foregroundStyle(colour)
                        .lineLimit(1)
                        .truncationMode(.tail)
                    if let state = shortState(watch.readiness, now: now) {
                        Text(state)
                            .font(.caption)
                            .foregroundStyle(SproutColor.onSurfaceVariant)
                            .lineLimit(1)
                            .layoutPriority(1)
                    }
                    Spacer(minLength: 0)
                }
                .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            // One utterance, and the long sentence rather than the short one:
            // the line is abbreviated because it is being *looked* at, and none
            // of that applies to VoiceOver.
            .accessibilityElement(children: .ignore)
            .accessibilityLabel(
                "\(watch.medicine.name). \(stateSentence(watch.readiness, now: now))"
            )

            // Never disabled, whatever the light says — the same rule the As
            // needed screen keeps. Sprout records what happened; it does not
            // decide it.
            Button(Str.t("medicine_give_short"), action: onGive)
                .font(.footnote.weight(.medium))
                .buttonStyle(.plain)
                .foregroundStyle(SproutColor.primary)

            Button(action: onDismiss) {
                Image(systemName: "xmark")
                    .font(.footnote)
                    .foregroundStyle(SproutColor.outline)
                    .padding(.leading, Spacing.tight)
            }
            .buttonStyle(.plain)
            .accessibilityLabel(Str.t("medicine_dismiss"))
        }
        .padding(.horizontal, Spacing.snug)
        .padding(.vertical, Spacing.tight)
    }
}
